# Matrizes e vetores

## License

All content can be freely used and adapted under the terms of the 
[Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/).

![Creative Commons License](https://i.creativecommons.org/l/by/4.0/88x31.png)

## Representação de uma matriz

Antes de podermos manipular matrizes e vetores no computador, precisamos de um jeito de armazená-los em variáveis.
Para vetores, o candidato natural são **listas**.

$$
\mathbf{v} = \begin{bmatrix}1 \\ 2 \\ 3\end{bmatrix}
$$

O vetor acima poderia ser representado em código como:



In [1]:
v = [1, 2, 3]
print(v)

[1, 2, 3]


Uma matriz pode ser vista como um conjunto de vetores, ou um vetor de vetores. Cada vetor seria equivalente a uma linha da matriz:

$$
\mathbf{A} =
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9 \\
\end{bmatrix} =
\begin{bmatrix}
[1 & 2 & 3] \\
[4 & 5 & 6] \\
[7 & 8 & 9] \\
\end{bmatrix}
$$

Logo, um jeito de representar uma matriz em Python é através de uma **lista de listas**:

In [2]:
A = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


O Python permite quebrar a linha quando o comando está entre `[` ou `(`:

In [3]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
print(A)

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]


*Note que o print acima imprime a matriz toda em uma única linha*.

A matriz `A` é uma lista como qualquer outra. A única diferença é que **os elementos da lista são outras listas**. O elemnto `[0]` da lista `A` será a lista correspondente a primeira linha:

In [4]:
print(A[0])

[1, 2, 3]


Como `A[0]` é uma lista, podemos pegar os elementos dessa lista da mesma forma:

In [5]:
print(A[0][0])
print(A[1][2])

1
6


### Exemplo

Queremos imprimir cada elemento da matriz `A` da seguinte forma:

    1  2  3  
    4  5  6  
    7  8  9

Podemos utilizar o `for` para pegar cada linha da matriz. Para cada linha da matriz, queremos imprimir o elemento. Quando terminarmos de imprimir a linha, vamos pular uma linha.

In [6]:
for i in range(3): # Anda sobre as linhas
    for j in range(3): # Anda sobre as colunas
        print(A[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime nada e pula uma linha

1  2  3  
4  5  6  
7  8  9  


## Somando matrizes

A soma de duas matrizes produz uma matriz que terá a soma dos elementos correspondentes:

$$
\begin{bmatrix}
a & b & c \\
d & e & f \\
g & h & i \\
\end{bmatrix} +
\begin{bmatrix}
j & l & m \\
n & o & p \\
q & r & s \\
\end{bmatrix} =
\begin{bmatrix}
a + j & b + l & c + m \\
d + n & e + o & f + p \\
g + q & h + r & i + s \\
\end{bmatrix}
$$

De forma genérica, o $j$-ésimo elemento da $i$-ésima linha de uma matriz $\mathbf{A}$ é $A_{ij}$.
Se temos duas matrizes $\mathbf{A}$ e $\mathbf{B}$, a soma pode ser escrita como:

$$
C_{ij} = A_{ij} + B_{ij}
$$

## Tarefa

Some as matrizes `A` e `B` definidas abaixo e guarde o resultado em uma matriz `C`. Imprima a matriz `C`.

**Dicas**:

* Você pode criar a matriz `C` antes ou durante a soma. Lembre-se que a matriz é uma lista e listas possuem o método `append`.

In [7]:
A = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11]]
nlin_a = 3
nlin_b = 3
ncol_a = 3
ncol_b = 3

In [8]:
c = []
for i in range (3):
    c.append([]) #cria uma linha na matriz soma
    for j in range (3):
        soma = A[i][j] + B[i][j] # soma é a variável para a soma das matrizes A e B
        c[i].append(soma)
        print(c[i][j], ' ', end='') # end ='' não pula as linhas no print
    print () # imprime vazio e pula uma linha

4  6  8  
10  12  14  
16  18  20  


In [10]:
print (c) # mostrando a diferença de fazer com o end="" e um print sem uma  

[[4, 6, 8], [10, 12, 14], [16, 18, 20]]


### Resultado esperado

Seu código deve imprimir exatamente:

    4  6  8  
    10  12  14  
    16  18  20  

## Multiplicando uma matriz por um vetor

A multiplicação de uma matriz por um vetor é

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e \\
f \\
\end{bmatrix} =
\begin{bmatrix}
ae + bf \\
ce + df \\
\end{bmatrix}
$$

Sendo o vetor $\mathbf{u} = \mathbf{A}\mathbf{v}$, cada elemento $i$ de $\mathbf{u}$ é 


$$
u_i = \sum\limits_{k=1}^{N} A_{ik}v_k
$$

## Tarefa

Faça a multiplicação da matriz pelo vetor definidos abaixo. Guarde o resultado em uma lista.

**Dicas**

In [11]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
v = [12, 13, 14, 15]
nlin = 3
ncol = 4

In [12]:
matriz_vetor = []
for i in range (nlin):
    
    multi_vet = 0
    for j in range(ncol):
        multi_vet = multi_vet + A[i][j] * v[j] #multiplicando o vetor pela matriz
       
    matriz_vetor.append(multi_vet) # valores da matriz guardados  
print (matriz_vetor)

[140, 302, 464]


### Resultado esperado

O seu código deve imprimir exatamente:

    [140, 302, 464]

## Multiplicação de matrizes

A multiplicação de matrizes é feita de forma diferente da soma. É mais fácil mostrar do que explicar:

$$
\begin{bmatrix}
a & b  \\
c & d  \\
\end{bmatrix} 
\begin{bmatrix}
e & f \\
g & h  \\
\end{bmatrix} =
\begin{bmatrix}
ae + bg & af + bh  \\
ce + dg & cf + dh  \\
\end{bmatrix}
$$

Sendo $\mathbf{C} = \mathbf{A}\mathbf{B}$, cada elemento $ij$ de $\mathbf{C}$ é 


$$
C_{ij} = \sum\limits_{k=1}^{N} A_{ik}B_{kj}
$$

## Tarefa

Faça a multiplicação das duas matrizes definidas abaixo. Guarde o resultado em uma matriz (lista de listas).

In [15]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
B = [[3, 4, 5],
     [6, 7, 8],
     [9, 10, 11],
     [12, 13, 14]]
nlin_a = 3
nlin_b = 4
ncol_a = 4
ncol_b = 3

In [26]:
D =[] #criando uma lista vazia 
for i in range (nlin_a): # Criaçao de um range para o numero de linhas de A
    mult =[] 
    for j in range(ncol_b): # criaçao de um range para o numero de colunas de B
        mult_1 = 0 # criação da variavel mult_1
        for k in range(ncol_a):#criação de um range para o número de colunas da matriz nova
            mult_1 = A[i][k]*B[k][j] + mult_1 # representando a multipicação das matrizes
        mult.append(mult_1) #append da multipicação em mult
    D.append(mult) #append da multipilicação em D
print (D) #imprime a matriz D gerada pela multpicação de A e B em linha

[[90, 100, 110], [180, 202, 224], [270, 304, 338]]


In [27]:
for i in range(3): # criaçao de um range  para o numero de linhas de D
    for j in range(3): # criaçao de um range para o numero de colunas de D
        print(D[i][j], " ", end=" ") # printa uma linha em baixo da outra
    print() # imprime vazio e pula uma linha

90   100   110   
180   202   224   
270   304   338   


### Resultado esperado

O seu código deve imprimir exatamente:

    90  100  110  
    180  202  224  
    270  304  338  



## Tarefa Bônus

Calcule o produto $\mathbf{A}^T\mathbf{A}$ para a matriz $A$ definida abaixo. $\mathbf{A}^T$ é a matriz transposta de $A$:

$$
\mathbf{A} = 
\begin{bmatrix}
a & b & c \\
d & e & f \\
\end{bmatrix}
$$

$$
\mathbf{A}^T = 
\begin{bmatrix}
a & d \\
b & e \\
c & f \\
\end{bmatrix}
$$

Ou seja, as linhas de $\mathbf{A}$ são as colunas de $\mathbf{A}^T$.


In [28]:
A = [[1, 2, 3, 4],
     [4, 5, 6, 7],
     [7, 8, 9, 10]]
nlin = 3
ncol = 4

In [29]:
transposta = [] #criando uma lista vazia
for j in range(ncol):
    transposta.append([]) #cria uma linha na matriz transposta
    for i in range(nlin):
        mudança = 0 #crianção de uma variável mudança
        mudança = A[i][j] #mundaça é a variável pra gerar a matriz transposta
        transposta[j].append(mudança) #append da matriz transposta em transposta
        print(A[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() # Imprime vazio e pula uma linha

1  4  7  
2  5  8  
3  6  9  
4  7  10  


In [30]:
nlin_t = 4 #número de linhas da matriz tranposta 
ncol_t = 3 #número de colunas da matriz transposta
print(transposta) #imprime os valores da matriz transposta em lista 

[[1, 4, 7], [2, 5, 8], [3, 6, 9], [4, 7, 10]]


In [31]:
M = [] # criação de uma lista vazia
for i in range(nlin_t): # representaçao do numero de linhas da matriz transposta
    math = [] 
    for j in range(ncol): # representaçao do numero de colunas da matriz A
        multi_mat = 0 #crianção da variável multi_mat
        for k in range(ncol_t): # representaçao do numero de colunas da matriz transposta
            multi_mat = transposta[i][k]*A[k][j] + multi_mat #multi_mat é a variável para a multiplicação das matrizes A e a sua transposta
        math.append(multi_mat) #append da multipicação em math 
    M.append(math) #append da multiplicação em M 
print(M) #imprime a matriz M gerada pela multpicação de A e a sua transposta em linha

[[66, 78, 90, 102], [78, 93, 108, 123], [90, 108, 126, 144], [102, 123, 144, 165]]


In [32]:
for i in range(4): # representaçao do numero de linhas da matriz M
    for j in range(4): # representaçao do numero de colunas da matriz M 
        print(M[i][j], ' ', end='') # end='' faz com que print não pule uma linha
    print() #imprime vazio e pula uma linha

66  78  90  102  
78  93  108  123  
90  108  126  144  
102  123  144  165  


### Resultado esperado

O seu código deve imprimir exatamente:

    66  78  90  102  
    78  93  108  123  
    90  108  126  144  
    102  123  144  165  
